target_chains/solana/programs/pyth-price-publisher/src/accounts/publisher_prices.rs

```diff
@@ -0,0 +1,173 @@
+//! The PublisherPrices account acts as the buffer for storing prices sent
+//! by publishers. It tracks the slot in which it is collecting in order
+//! to allow the validator to stay in sync.
+
+use {
+    bytemuck::{cast_slice, from_bytes, from_bytes_mut, Pod, Zeroable},
+    solana_program::clock::Slot,
+    std::mem::size_of,
+    thiserror::Error,
+};
+
+/// Account Magic to avoid Account Confusiong
+const FORMAT: u32 = 2848712303;
+
+#[derive(Debug, Clone, Copy, Zeroable, Pod)]
+#[repr(C, packed)]
+pub struct PublisherPricesHeader {
+    pub format: u32,
+    pub publisher: [u8; 32],
+    pub slot: Slot,
+    pub num_prices: u32,
+}
+
+impl PublisherPricesHeader {
+    pub fn new(publisher: [u8; 32]) -> Self {
+        PublisherPricesHeader {
+            format: FORMAT,
+            publisher,
+            slot: 0,
+            num_prices: 0,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Zeroable, Pod)]
+#[repr(C, packed)]
+pub struct PublisherPrice {
+    // 4 high bits: trading status
+    // 28 low bits: feed index
+    pub trading_status_and_feed_index: u32,
+    pub price: i64,
+    pub confidence: u64,
+}
+
+#[derive(Debug, Error)]
+#[error("publisher price data overflow")]
+pub struct PublisherPriceError;
+
+impl PublisherPrice {
+    pub fn new(
+        feed_index: u32,
+        trading_status: u32,
+        price: i64,
+        confidence: u64,
+    ) -> Result<Self, PublisherPriceError> {
+        if feed_index >= (1 << 28) || trading_status >= (1 << 4) {
+            return Err(PublisherPriceError);
+        }
+        Ok(Self {
+            trading_status_and_feed_index: (trading_status << 28) | feed_index,
+            price,
+            confidence,
+        })
+    }
+
+    pub fn trading_status(&self) -> u32 {
+        self.trading_status_and_feed_index >> 28
+    }
+
+    pub fn feed_index(&self) -> u32 {
+        self.trading_status_and_feed_index & ((1 << 28) - 1)
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum ReadAccountError {
+    #[error("data too short")]
+    DataTooShort,
+    #[error("format mismatch")]
+    FormatMismatch,
+    #[error("invalid num prices")]
+    InvalidNumPrices,
+}
+
+#[derive(Debug, Error)]
+pub enum ExtendError {
+    #[error("not enough space")]
+    NotEnoughSpace,
+    #[error("invalid length")]
+    InvalidLength,
+}
+
+pub fn format_matches(data: &[u8]) -> bool {
+    if data.len() < size_of::<u32>() {
+        return false;
+    }
+    let format: &u32 = from_bytes(&data[..size_of::<u32>()]);
+    *format == FORMAT
+}
+
+pub fn read(data: &[u8]) -> Result<(&PublisherPricesHeader, &[PublisherPrice]), ReadAccountError> {
```

Comment:
it's a bit strange that read and read_mut have different signatures.

Reply:
It's a bit awkward because I want `read` to return only the initialized part of the account as `&[PublisherPrice]` slice, but in `read_mut` I need to return a full slice of the remainder of the account so that we can call `extend` later. To make it more clear that not all values returned by `read_mut` are properly initialized `PublisherPrice` values, I decided to return `&mut [u8]` instead of `&mut [PublisherPrice]`. I thought about encapsulating it in a new struct like this:
```
let mut extender = read_mut(data);
let publisher = extender.publisher();
extender.extend(new_prices);
```
But I decided to keep it simple for now. Let me know your thoughts.